src/services/exportInfoMap.ts

```diff
@@ -445,15 +447,30 @@ export function forEachExternalModuleToImportFrom(
     }
 }
 
-function forEachExternalModule(checker: TypeChecker, allSourceFiles: readonly SourceFile[], excludePatterns: readonly RegExp[] | undefined, cb: (module: Symbol, sourceFile: SourceFile | undefined) => void) {
-    const isExcluded = excludePatterns && ((fileName: string) => excludePatterns.some(p => p.test(fileName)));
+function forEachExternalModule(checker: TypeChecker, allSourceFiles: readonly SourceFile[], excludePatterns: readonly RegExp[] | undefined, host: LanguageServiceHost, cb: (module: Symbol, sourceFile: SourceFile | undefined) => void) {
+    const realpathsWithSymlinks = host.getSymlinkCache?.().getSymlinkedDirectoriesByRealpath();
+    const isExcluded = excludePatterns && (({ fileName, path }: SourceFile) => {
+        if (excludePatterns.some(p => p.test(fileName))) return true;
+        if (realpathsWithSymlinks?.size && pathContainsNodeModules(fileName)) {
+            let dir = getDirectoryPath(fileName);

```

Comment:
Just so I'm clear, this is needed because we're basically doing `forEachAncestorDirectory` on two paths at once?

Could `dir` bottom out before `path` does?

Reply:
`dirPath === toPath(dir)` (i.e., `dir` has original casing preserved, but we need a `Path` for cache lookups). So no, they’ll always have the same number of path components.

Reply:
Oh, I think I'm just getting confused because this code uses both `fileName` and `path` off of SourceFile, which are the same thing but with different types?

Reply:
`path` is normalized to lowercase on case-insensitive file systems, whereas `fileName` retains the casing as it exists on disk. I’m not sure it actually matters here because the exclude regex is also case-insensitive on case-insensitive file systems, so I _think_ we could just use `path` and it would always work, but it felt more correct to check the pattern against the original filename, not the `Path`-ified one.

Reply:
Oh, thanks. I legitimately didn't know.